File: spicedb-go/client/lookup_test.go
package client

import (
	"context"
	"net"
	"testing"

	"github.com/stretchr/testify/require"
	"google.golang.org/grpc"
	"google.golang.org/grpc/test/bufconn"

	v1 "github.com/authzed/spicedb-clients/proto-clients/spicedb-go-proto/gen/authzed/api/v1"
	"github.com/authzed/spicedb-clients/spicedb-go/consistency"
)

// lookupStubServer is a stub PermissionsService whose LookupResources and
// LookupSubjects implementations stream fixed, synthetic responses so tests
// can assert on how the client maps proto responses to native result types.
type lookupStubServer struct {
	v1.UnimplementedPermissionsServiceServer

	lookupResourcesResponses []*v1.LookupResourcesResponse
	lookupSubjectsResponses  []*v1.LookupSubjectsResponse
}

func (s *lookupStubServer) LookupResources(_ *v1.LookupResourcesRequest, stream grpc.ServerStreamingServer[v1.LookupResourcesResponse]) error {
	for _, resp := range s.lookupResourcesResponses {
		if err := stream.Send(resp); err != nil {
			return err
		}
	}
	return nil
}

func (s *lookupStubServer) LookupSubjects(_ *v1.LookupSubjectsRequest, stream grpc.ServerStreamingServer[v1.LookupSubjectsResponse]) error {
	for _, resp := range s.lookupSubjectsResponses {
		if err := stream.Send(resp); err != nil {
			return err
		}
	}
	return nil
}

// startLookupStubServer starts an in-process gRPC server backed by bufconn,
// exposing a PermissionsService backed by the given stub, and returns a
// dialer for it.
func startLookupStubServer(t *testing.T, stub *lookupStubServer) func(context.Context, string) (net.Conn, error) {
	t.Helper()

	const bufSize = 1024 * 1024
	lis := bufconn.Listen(bufSize)

	grpcServer := grpc.NewServer()
	v1.RegisterPermissionsServiceServer(grpcServer, stub)

	go func() {
		_ = grpcServer.Serve(lis)
	}()
	t.Cleanup(grpcServer.Stop)

	return func(ctx context.Context, _ string) (net.Conn, error) {
		return lis.DialContext(ctx)
	}
}

// newTestClient dials the given stub server over bufconn and returns a
// ready-to-use *Client.
func newTestClient(t *testing.T, dialer func(context.Context, string) (net.Conn, error)) *Client {
	t.Helper()

	c, err := NewWithOpts("passthrough:///bufnet", "test-token",
		WithInsecure(),
		WithDialOptions(grpc.WithContextDialer(dialer)),
	)
	require.NoError(t, err)
	return c
}

// TestLookupResources_YieldsPermissionshipAndPartialCaveat proves that
// LookupResources surfaces the proto's permissionship and partial caveat
// info instead of dropping them, so a CONDITIONAL match is distinguishable
// from a full HAS_PERMISSION grant.
func TestLookupResources_YieldsPermissionshipAndPartialCaveat(t *testing.T) {
	stub := &lookupStubServer{
		lookupResourcesResponses: []*v1.LookupResourcesResponse{
			{
				ResourceObjectId: "doc1",
				Permissionship:   v1.LookupPermissionship_LOOKUP_PERMISSIONSHIP_HAS_PERMISSION,
			},
			{
				ResourceObjectId: "doc2",
				Permissionship:   v1.LookupPermissionship_LOOKUP_PERMISSIONSHIP_CONDITIONAL_PERMISSION,
				PartialCaveatInfo: &v1.PartialCaveatInfo{
					MissingRequiredContext: []string{"ip_address"},
				},
			},
		},
	}
	dialer := startLookupStubServer(t, stub)
	c := newTestClient(t, dialer)

	var got []LookupResource
	for r, err := range c.LookupResources(context.Background(), consistency.MinLatency(), "document", "view", "user", "alice") {
		require.NoError(t, err)
		got = append(got, r)
	}

	require.Len(t, got, 2)

	require.Equal(t, "doc1", got[0].ResourceID)
	require.Equal(t, PermissionshipHasPermission, got[0].Permissionship)
	require.Nil(t, got[0].PartialCaveat)

	require.Equal(t, "doc2", got[1].ResourceID)
	require.Equal(t, PermissionshipConditionalPermission, got[1].Permissionship)
	require.NotNil(t, got[1].PartialCaveat)
	require.Equal(t, []string{"ip_address"}, got[1].PartialCaveat.MissingRequiredContext)
}

// TestLookupResources_StreamErrorYieldsZeroValueAndMappedError proves that a
// mid-stream error yields a zero-value LookupResource paired with a native
// mapped error, matching the (result, error) iterator contract used
// elsewhere in the client.
func TestLookupResources_StreamErrorYieldsZeroValueAndMappedError(t *testing.T) {
	// startErroringServer (from errors_test.go) only wires up ReadRelationships;
	// LookupResources hits UnimplementedPermissionsServiceServer, which is
	// enough to prove errors flow through mapGRPCError rather than panicking
	// or being silently swallowed.
	dialer := startErroringServer(t)
	c := newTestClient(t, dialer)

	var (
		yields int
		gotErr error
	)
	for r, err := range c.LookupResources(context.Background(), consistency.MinLatency(), "document", "view", "user", "alice") {
		yields++
		require.Equal(t, LookupResource{}, r)
		gotErr = err
	}

	require.Equal(t, 1, yields)
	require.Error(t, gotErr)
}

// TestLookupSubjects_WildcardSubjectExposesExcludedSubjects is the key
// over-grant-fix assertion: when LookupSubjects resolves a wildcard "*"
// subject, the excluded_subjects the server attaches to that wildcard MUST
// be surfaced to the caller, since dropping them would make a wildcard match
// look like an unconditional grant to every subject including the excluded
// ones.
func TestLookupSubjects_WildcardSubjectExposesExcludedSubjects(t *testing.T) {
	stub := &lookupStubServer{
		lookupSubjectsResponses: []*v1.LookupSubjectsResponse{
			{
				Subject: &v1.ResolvedSubject{
					SubjectObjectId: "*",
					Permissionship:  v1.LookupPermissionship_LOOKUP_PERMISSIONSHIP_HAS_PERMISSION,
				},
				ExcludedSubjects: []*v1.ResolvedSubject{
					{
						SubjectObjectId: "eve",
						Permissionship:  v1.LookupPermissionship_LOOKUP_PERMISSIONSHIP_HAS_PERMISSION,
					},
					{
						SubjectObjectId: "mallory",
						Permissionship:  v1.LookupPermissionship_LOOKUP_PERMISSIONSHIP_HAS_PERMISSION,
					},
				},
			},
		},
	}
	dialer := startLookupStubServer(t, stub)
	c := newTestClient(t, dialer)

	var got []LookupSubject
	for s, err := range c.LookupSubjects(context.Background(), consistency.MinLatency(), "document", "doc1", "view", "user") {
		require.NoError(t, err)
		got = append(got, s)
	}

	require.Len(t, got, 1)
	require.Equal(t, "*", got[0].Subject.SubjectID)
	require.Equal(t, PermissionshipHasPermission, got[0].Subject.Permissionship)

	require.Len(t, got[0].ExcludedSubjects, 2)
	excludedIDs := []string{got[0].ExcludedSubjects[0].SubjectID, got[0].ExcludedSubjects[1].SubjectID}
	require.ElementsMatch(t, []string{"eve", "mallory"}, excludedIDs)
}

// TestLookupSubjects_NonWildcardHasNoExcludedSubjects proves that a plain
// (non-wildcard) match doesn't spuriously populate ExcludedSubjects.
func TestLookupSubjects_NonWildcardHasNoExcludedSubjects(t *testing.T) {
	stub := &lookupStubServer{
		lookupSubjectsResponses: []*v1.LookupSubjectsResponse{
			{
				Subject: &v1.ResolvedSubject{
					SubjectObjectId: "alice",
					Permissionship:  v1.LookupPermissionship_LOOKUP_PERMISSIONSHIP_HAS_PERMISSION,
				},
			},
		},
	}
	dialer := startLookupStubServer(t, stub)
	c := newTestClient(t, dialer)

	var got []LookupSubject
	for s, err := range c.LookupSubjects(context.Background(), consistency.MinLatency(), "document", "doc1", "view", "user") {
		require.NoError(t, err)
		got = append(got, s)
	}

	require.Len(t, got, 1)
	require.Equal(t, "alice", got[0].Subject.SubjectID)
	require.Empty(t, got[0].ExcludedSubjects)
}

// TestLookupSubjects_FallsBackToDeprecatedFields proves that when a server
// only populates the deprecated top-level subject_object_id (and leaves the
// non-deprecated `subject` field unset), the client still surfaces a usable
// ResolvedSubject rather than an empty SubjectID.
func TestLookupSubjects_FallsBackToDeprecatedFields(t *testing.T) {
	stub := &lookupStubServer{
		lookupSubjectsResponses: []*v1.LookupSubjectsResponse{
			{
				SubjectObjectId: "bob",                                                        //nolint:staticcheck
				Permissionship:  v1.LookupPermissionship_LOOKUP_PERMISSIONSHIP_HAS_PERMISSION, //nolint:staticcheck
			},
		},
	}
	dialer := startLookupStubServer(t, stub)
	c := newTestClient(t, dialer)

	var got []LookupSubject
	for s, err := range c.LookupSubjects(context.Background(), consistency.MinLatency(), "document", "doc1", "view", "user") {
		require.NoError(t, err)
		got = append(got, s)
	}

	require.Len(t, got, 1)
	require.Equal(t, "bob", got[0].Subject.SubjectID)
	require.Equal(t, PermissionshipHasPermission, got[0].Subject.Permissionship)
}

// TestLookupSubjects_ExcludedSubjectsFallsBackToDeprecatedIds proves that
// when a server populates a wildcard "*" match's exclusions ONLY via the
// deprecated top-level excluded_subject_ids field (leaving the
// non-deprecated excluded_subjects list empty), the client still surfaces
// those exclusions as ResolvedSubjects. This is security-relevant: dropping
// exclusions from an older-wire-format server would silently over-grant
// access to the excluded subjects. Removing the excluded_subject_ids
// fallback branch in LookupSubjects MUST fail this test.
func TestLookupSubjects_ExcludedSubjectsFallsBackToDeprecatedIds(t *testing.T) {
	stub := &lookupStubServer{
		lookupSubjectsResponses: []*v1.LookupSubjectsResponse{
			{
				Subject: &v1.ResolvedSubject{
					SubjectObjectId: "*",
					Permissionship:  v1.LookupPermissionship_LOOKUP_PERMISSIONSHIP_HAS_PERMISSION,
				},
				// Non-deprecated excluded_subjects deliberately left empty;
				// only the deprecated excluded_subject_ids is populated.
				ExcludedSubjectIds: []string{"eve", "mallory"}, //nolint:staticcheck
			},
		},
	}
	dialer := startLookupStubServer(t, stub)
	c := newTestClient(t, dialer)

	var got []LookupSubject
	for s, err := range c.LookupSubjects(context.Background(), consistency.MinLatency(), "document", "doc1", "view", "user") {
		require.NoError(t, err)
		got = append(got, s)
	}

	require.Len(t, got, 1)
	require.Equal(t, "*", got[0].Subject.SubjectID)
	require.Len(t, got[0].ExcludedSubjects, 2)
	require.Equal(t, []ResolvedSubject{
		{SubjectID: "eve"},
		{SubjectID: "mallory"},
	}, got[0].ExcludedSubjects)
}

// TestLookupSubjects_StreamErrorYieldsZeroValueAndMappedError proves the
// (result, error) iterator contract holds for LookupSubjects too.
func TestLookupSubjects_StreamErrorYieldsZeroValueAndMappedError(t *testing.T) {
	dialer := startErroringServer(t)
	c := newTestClient(t, dialer)

	var (
		yields int
		gotErr error
	)
	for s, err := range c.LookupSubjects(context.Background(), consistency.MinLatency(), "document", "doc1", "view", "user") {
		yields++
		require.Equal(t, LookupSubject{}, s)
		gotErr = err
	}

	require.Equal(t, 1, yields)
	require.Error(t, gotErr)
}

File: spicedb-typescript/src/__tests__/lookup.test.ts
import { describe, it, expect } from "vitest";
import { create } from "@bufbuild/protobuf";
import {
  LookupResourcesResponseSchema,
  LookupSubjectsResponseSchema,
  ResolvedSubjectSchema,
  PartialCaveatInfoSchema,
  LookupPermissionship,
} from "@spicedb/proto";
import {
  permissionshipFromProto,
  partialCaveatFromProto,
  resolvedSubjectFromProto,
  fromProtoLookupResource,
  fromProtoLookupSubject,
} from "../types.js";

describe("permissionshipFromProto()", () => {
  it("maps HAS_PERMISSION", () => {
    expect(
      permissionshipFromProto(LookupPermissionship.HAS_PERMISSION),
    ).toBe("hasPermission");
  });

  it("maps CONDITIONAL_PERMISSION", () => {
    expect(
      permissionshipFromProto(LookupPermissionship.CONDITIONAL_PERMISSION),
    ).toBe("conditionalPermission");
  });

  it("maps UNSPECIFIED", () => {
    expect(permissionshipFromProto(LookupPermissionship.UNSPECIFIED)).toBe(
      "unspecified",
    );
  });
});

describe("partialCaveatFromProto()", () => {
  it("maps undefined to undefined", () => {
    expect(partialCaveatFromProto(undefined)).toBeUndefined();
  });

  it("maps a present PartialCaveatInfo", () => {
    const proto = create(PartialCaveatInfoSchema, {
      missingRequiredContext: ["region", "time"],
    });
    expect(partialCaveatFromProto(proto)).toEqual({
      missingRequiredContext: ["region", "time"],
    });
  });
});

describe("resolvedSubjectFromProto()", () => {
  it("maps subjectObjectId, permissionship, and partialCaveat", () => {
    const proto = create(ResolvedSubjectSchema, {
      subjectObjectId: "sally",
      permissionship: LookupPermissionship.CONDITIONAL_PERMISSION,
      partialCaveatInfo: create(PartialCaveatInfoSchema, {
        missingRequiredContext: ["region"],
      }),
    });
    expect(resolvedSubjectFromProto(proto)).toEqual({
      subjectId: "sally",
      permissionship: "conditionalPermission",
      partialCaveat: { missingRequiredContext: ["region"] },
    });
  });

  it("maps an undefined input to a zero-value ResolvedSubject", () => {
    expect(resolvedSubjectFromProto(undefined)).toEqual({
      subjectId: "",
      permissionship: "unspecified",
      partialCaveat: undefined,
    });
  });
});

describe("fromProtoLookupResource()", () => {
  it("maps a HAS_PERMISSION result with no partial caveat", () => {
    const resp = create(LookupResourcesResponseSchema, {
      resourceObjectId: "readme",
      permissionship: LookupPermissionship.HAS_PERMISSION,
    });
    expect(fromProtoLookupResource(resp)).toEqual({
      resourceId: "readme",
      permissionship: "hasPermission",
      partialCaveat: undefined,
    });
  });

  it("maps a CONDITIONAL_PERMISSION result with its partial caveat", () => {
    const resp = create(LookupResourcesResponseSchema, {
      resourceObjectId: "design",
      permissionship: LookupPermissionship.CONDITIONAL_PERMISSION,
      partialCaveatInfo: create(PartialCaveatInfoSchema, {
        missingRequiredContext: ["is_in_region"],
      }),
    });
    expect(fromProtoLookupResource(resp)).toEqual({
      resourceId: "design",
      permissionship: "conditionalPermission",
      partialCaveat: { missingRequiredContext: ["is_in_region"] },
    });
  });
});

describe("fromProtoLookupSubject()", () => {
  it("surfaces a wildcard subject with two excluded subjects (modern fields)", () => {
    const resp = create(LookupSubjectsResponseSchema, {
      subject: create(ResolvedSubjectSchema, {
        subjectObjectId: "*",
        permissionship: LookupPermissionship.HAS_PERMISSION,
      }),
      excludedSubjects: [
        create(ResolvedSubjectSchema, {
          subjectObjectId: "eve",
          permissionship: LookupPermissionship.HAS_PERMISSION,
        }),
        create(ResolvedSubjectSchema, {
          subjectObjectId: "mallory",
          permissionship: LookupPermissionship.HAS_PERMISSION,
        }),
      ],
    });
    const result = fromProtoLookupSubject(resp);
    expect(result.subject.subjectId).toBe("*");
    expect(result.excludedSubjects).toHaveLength(2);
    expect(result.excludedSubjects.map((s) => s.subjectId)).toEqual([
      "eve",
      "mallory",
    ]);
  });

  it("distinguishes HAS_PERMISSION vs CONDITIONAL_PERMISSION on the subject", () => {
    const resp = create(LookupSubjectsResponseSchema, {
      subject: create(ResolvedSubjectSchema, {
        subjectObjectId: "jimmy",
        permissionship: LookupPermissionship.CONDITIONAL_PERMISSION,
        partialCaveatInfo: create(PartialCaveatInfoSchema, {
          missingRequiredContext: ["region"],
        }),
      }),
    });
    const result = fromProtoLookupSubject(resp);
    expect(result.subject).toEqual({
      subjectId: "jimmy",
      permissionship: "conditionalPermission",
      partialCaveat: { missingRequiredContext: ["region"] },
    });
  });

  it("falls back to deprecated subjectObjectId when subject is unset", () => {
    const resp = create(LookupSubjectsResponseSchema, {
      subjectObjectId: "sally",
      permissionship: LookupPermissionship.HAS_PERMISSION,
    });
    const result = fromProtoLookupSubject(resp);
    expect(result.subject).toEqual({
      subjectId: "sally",
      permissionship: "hasPermission",
      partialCaveat: undefined,
    });
  });

  it("falls back to deprecated excludedSubjectIds when excludedSubjects is empty", () => {
    const resp = create(LookupSubjectsResponseSchema, {
      subjectObjectId: "*",
      permissionship: LookupPermissionship.HAS_PERMISSION,
      excludedSubjectIds: ["eve", "mallory"],
    });
    const result = fromProtoLookupSubject(resp);
    expect(result.subject.subjectId).toBe("*");
    expect(result.excludedSubjects).toEqual([
      { subjectId: "eve", permissionship: "unspecified", partialCaveat: undefined },
      { subjectId: "mallory", permissionship: "unspecified", partialCaveat: undefined },
    ]);
  });

  it("prefers modern excludedSubjects over deprecated excludedSubjectIds when both are present", () => {
    const resp = create(LookupSubjectsResponseSchema, {
      subject: create(ResolvedSubjectSchema, { subjectObjectId: "*" }),
      excludedSubjects: [
        create(ResolvedSubjectSchema, {
          subjectObjectId: "eve",
          permissionship: LookupPermissionship.HAS_PERMISSION,
        }),
      ],
      excludedSubjectIds: ["stale-legacy-id"],
    });
    const result = fromProtoLookupSubject(resp);
    expect(result.excludedSubjects).toEqual([
      { subjectId: "eve", permissionship: "hasPermission", partialCaveat: undefined },
    ]);
  });
});
